File: web/src/store/applyOperation.ts
import type { OperationType } from '../types/collaboration'
import type { Attribute, CanonicalModel, ClassEntity, Method, Relationship, Waypoint } from '../types/diagram'

/**
 * Espejo en TypeScript de CanonicalModelMutator (backend): aplica sobre una copia
 * en memoria del CanonicalModel el mismo catálogo de operaciones de la sección 8.1,
 * usado para reconstruir el estado local a partir de los StompBroadcastMessage
 * recibidos en /topic/diagrams/{id}. Solo cubre las operaciones que este cliente
 * puede emitir o necesita reflejar (ver alcance de la Fase 1: sin paquetes,
 * waypoints, bulk-merge ni cursores).
 */
export function applyOperation(
  model: CanonicalModel,
  operationType: OperationType,
  targetId: string | null,
  payload: Record<string, unknown>,
): CanonicalModel {
  switch (operationType) {
    case 'ADD_CLASS':
      return addClass(model, payload)
    case 'MOVE_CLASS':
      return updateClassById(model, targetId, (c) => ({
        ...c,
        position: { x: numberField(payload, 'x', c.position.x), y: numberField(payload, 'y', c.position.y) },
      }))
    case 'RESIZE_CLASS':
      return updateClassById(model, targetId, (c) => ({
        ...c,
        width: numberField(payload, 'width', c.width),
        height: numberField(payload, 'height', c.height),
      }))
    case 'RENAME_CLASS':
      return updateClassById(model, targetId, (c) => ({
        ...c,
        name: stringField(payload, 'name') ?? c.name,
      }))
    case 'DELETE_CLASS':
      return {
        ...model,
        classes: model.classes.filter((c) => c.id !== targetId),
        relationships: model.relationships.filter(
          (r) => r.sourceClassId !== targetId && r.targetClassId !== targetId,
        ),
      }
    case 'ADD_ATTRIBUTE':
      return updateClassById(model, targetId, (c) => {
        const newAttribute = payload as unknown as Attribute
        if (c.attributes.some((a) => a.id === newAttribute.id)) return c
        return { ...c, attributes: [...c.attributes, newAttribute] }
      })
    case 'UPDATE_ATTRIBUTE':
      return updateAttributeById(model, targetId, (a) => ({ ...a, ...payload }))
    case 'DELETE_ATTRIBUTE':
      return {
        ...model,
        classes: model.classes.map((c) => ({
          ...c,
          attributes: c.attributes.filter((a) => a.id !== targetId),
        })),
      }
    case 'ADD_METHOD':
      return updateClassById(model, targetId, (c) => {
        const newMethod = payload as unknown as Method
        if (c.methods.some((m) => m.id === newMethod.id)) return c
        return { ...c, methods: [...c.methods, newMethod] }
      })
    case 'UPDATE_METHOD':
      return updateMethodById(model, targetId, (m) => ({ ...m, ...payload }))
    case 'DELETE_METHOD':
      return {
        ...model,
        classes: model.classes.map((c) => ({
          ...c,
          methods: c.methods.filter((m) => m.id !== targetId),
        })),
      }
    case 'ADD_RELATIONSHIP': {
      const newRelationship = payload as unknown as Relationship
      if (model.relationships.some((r) => r.id === newRelationship.id)) return model
      return { ...model, relationships: [...model.relationships, newRelationship] }
    }
    case 'DELETE_RELATIONSHIP':
      return { ...model, relationships: model.relationships.filter((r) => r.id !== targetId) }
    case 'UPDATE_WAYPOINTS': {
      // A diferencia de UPDATE_RELATIONSHIP (que preserva los waypoints existentes
      // a propósito, ver CanonicalModelMutator.updateRelationship), esta es la ÚNICA
      // operación que los modifica -- LWW_FREE, igual que MOVE_CLASS, para arrastre fluido.
      const waypoints = (payload.waypoints as Waypoint[] | undefined) ?? []
      return {
        ...model,
        relationships: model.relationships.map((r) => (r.id === targetId ? { ...r, waypoints } : r)),
      }
    }
    case 'UPDATE_RELATIONSHIP': {
      // Solo type/owningSide/joinTableName/multiplicidades/roles/isNavigable son mutables
      // (id, sourceClassId, targetClassId y waypoints se preservan siempre, igual que en
      // CanonicalModelMutator.updateRelationship del backend, para no "destruir y recrear
      // el conector").
      return {
        ...model,
        relationships: model.relationships.map((r) => (r.id === targetId ? { ...r, ...payload } : r)),
      }
    }
    default:
      // ACQUIRE_LOCK/RELEASE_LOCK/paquetes/BULK_MERGE/USER_CURSOR:
      // fuera de alcance de este núcleo (sin locks, sin paquetes, sin cursores todavía).
      return model
  }
}

function addClass(model: CanonicalModel, payload: Record<string, unknown>): CanonicalModel {
  const newClass = payload as unknown as ClassEntity
  if (model.classes.some((c) => c.id === newClass.id)) {
    return model
  }
  return { ...model, classes: [...model.classes, newClass] }
}

function updateClassById(
  model: CanonicalModel,
  classId: string | null,
  updater: (classEntity: ClassEntity) => ClassEntity,
): CanonicalModel {
  if (!classId) return model
  return {
    ...model,
    classes: model.classes.map((c) => (c.id === classId ? updater(c) : c)),
  }
}

function updateAttributeById(
  model: CanonicalModel,
  attributeId: string | null,
  updater: (attribute: Attribute) => Attribute,
): CanonicalModel {
  if (!attributeId) return model
  return {
    ...model,
    classes: model.classes.map((c) => ({
      ...c,
      attributes: c.attributes.map((a) => (a.id === attributeId ? updater(a) : a)),
    })),
  }
}

function updateMethodById(
  model: CanonicalModel,
  methodId: string | null,
  updater: (method: Method) => Method,
): CanonicalModel {
  if (!methodId) return model
  return {
    ...model,
    classes: model.classes.map((c) => ({
      ...c,
      methods: c.methods.map((m) => (m.id === methodId ? updater(m) : m)),
    })),
  }
}

function stringField(payload: Record<string, unknown>, key: string): string | null {
  const v = payload[key]
  return typeof v === 'string' ? v : null
}

function numberField(payload: Record<string, unknown>, key: string, fallback: number): number {
  const v = payload[key]
  return typeof v === 'number' ? v : fallback
}

File: movil/lib/features/diagram/apply_operation.dart
import 'diagram_models.dart';

/// Espejo en Dart de `web/src/store/applyOperation.ts`, que a su vez espeja
/// `CanonicalModelMutator` (backend): aplica sobre una copia en memoria del
/// [CanonicalModel] el mismo catálogo de operaciones STOMP (sección 8.1),
/// reconstruyendo el estado local a partir de los `StompBroadcastMessage`
/// recibidos en `/topic/diagrams/{id}`. Solo lectura: no envía nada de vuelta.
CanonicalModel applyOperation(
  CanonicalModel model,
  String operationType,
  String? targetId,
  Map<String, dynamic> payload,
) {
  switch (operationType) {
    case 'ADD_CLASS':
      return _addClass(model, payload);
    case 'MOVE_CLASS':
      return _updateClassById(
        model,
        targetId,
        (c) => c.copyWith(
          position: Position(
            x: _numberField(payload, 'x', c.position.x),
            y: _numberField(payload, 'y', c.position.y),
          ),
        ),
      );
    case 'RESIZE_CLASS':
      return _updateClassById(
        model,
        targetId,
        (c) => c.copyWith(
          width: _numberField(payload, 'width', c.width),
          height: _numberField(payload, 'height', c.height),
        ),
      );
    case 'RENAME_CLASS':
      return _updateClassById(
        model,
        targetId,
        (c) => c.copyWith(name: payload['name'] as String? ?? c.name),
      );
    case 'DELETE_CLASS':
      return model.copyWith(
        classes: model.classes.where((c) => c.id != targetId).toList(),
        relationships: model.relationships
            .where((r) => r.sourceClassId != targetId && r.targetClassId != targetId)
            .toList(),
      );
    case 'ADD_ATTRIBUTE':
      return _updateClassById(model, targetId, (c) {
        final newAttribute = Attribute.fromJson(payload);
        if (c.attributes.any((a) => a.id == newAttribute.id)) return c;
        return c.copyWith(attributes: [...c.attributes, newAttribute]);
      });
    case 'UPDATE_ATTRIBUTE':
      return _updateAttributeById(model, targetId, (a) => a.mergeJson(payload));
    case 'DELETE_ATTRIBUTE':
      return model.copyWith(
        classes: model.classes
            .map((c) => c.copyWith(attributes: c.attributes.where((a) => a.id != targetId).toList()))
            .toList(),
      );
    case 'ADD_METHOD':
      return _updateClassById(model, targetId, (c) {
        final newMethod = Method.fromJson(payload);
        if (c.methods.any((m) => m.id == newMethod.id)) return c;
        return c.copyWith(methods: [...c.methods, newMethod]);
      });
    case 'UPDATE_METHOD':
      return _updateMethodById(model, targetId, (m) => m.mergeJson(payload));
    case 'DELETE_METHOD':
      return model.copyWith(
        classes: model.classes
            .map((c) => c.copyWith(methods: c.methods.where((m) => m.id != targetId).toList()))
            .toList(),
      );
    case 'ADD_RELATIONSHIP':
      final newRelationship = Relationship.fromJson(payload);
      if (model.relationships.any((r) => r.id == newRelationship.id)) return model;
      return model.copyWith(relationships: [...model.relationships, newRelationship]);
    case 'DELETE_RELATIONSHIP':
      return model.copyWith(relationships: model.relationships.where((r) => r.id != targetId).toList());
    case 'UPDATE_WAYPOINTS':
      final waypoints = (payload['waypoints'] as List<dynamic>? ?? [])
          .map((w) => Waypoint.fromJson(w as Map<String, dynamic>))
          .toList();
      return model.copyWith(
        relationships:
            model.relationships.map((r) => r.id == targetId ? r.copyWith(waypoints: waypoints) : r).toList(),
      );
    case 'UPDATE_RELATIONSHIP':
      return model.copyWith(
        relationships: model.relationships.map((r) => r.id == targetId ? r.mergeJson(payload) : r).toList(),
      );
    default:
      // ACQUIRE_LOCK/RELEASE_LOCK/paquetes/BULK_MERGE/USER_CURSOR: fuera de
      // alcance de este visor de solo lectura (sin locks, sin paquetes, sin
      // cursores todavía — mismo alcance que la Fase 1 del store web).
      return model;
  }
}

CanonicalModel _addClass(CanonicalModel model, Map<String, dynamic> payload) {
  final newClass = ClassEntity.fromJson(payload);
  if (model.classes.any((c) => c.id == newClass.id)) return model;
  return model.copyWith(classes: [...model.classes, newClass]);
}

CanonicalModel _updateClassById(
  CanonicalModel model,
  String? classId,
  ClassEntity Function(ClassEntity) updater,
) {
  if (classId == null) return model;
  return model.copyWith(
    classes: model.classes.map((c) => c.id == classId ? updater(c) : c).toList(),
  );
}

CanonicalModel _updateAttributeById(
  CanonicalModel model,
  String? attributeId,
  Attribute Function(Attribute) updater,
) {
  if (attributeId == null) return model;
  return model.copyWith(
    classes: model.classes
        .map((c) => c.copyWith(
              attributes: c.attributes.map((a) => a.id == attributeId ? updater(a) : a).toList(),
            ))
        .toList(),
  );
}

CanonicalModel _updateMethodById(
  CanonicalModel model,
  String? methodId,
  Method Function(Method) updater,
) {
  if (methodId == null) return model;
  return model.copyWith(
    classes: model.classes
        .map((c) => c.copyWith(
              methods: c.methods.map((m) => m.id == methodId ? updater(m) : m).toList(),
            ))
        .toList(),
  );
}

double _numberField(Map<String, dynamic> payload, String key, double fallback) {
  final value = payload[key];
  return value is num ? value.toDouble() : fallback;
}
